OIDC - Unable to recover username for auto onboard

With auto-onboard enabled, user login through OIDC fails with error in harbor-core: 
```
{"code":"INTERNAL_SERVER_ERROR","message":"unable to recover username for auto onboard, username claim: email"}
```
However, the username claim `email` **is** included in the user info string:
```
[DEBUG] [/core/controllers/oidc.go:225]: User info string: {"iss":"https://login.hooli.com/adfs","sub":"YzI5dFpWUm9hVzVuVW1GdVJHOXRSazlTWjJsVVNWTlRkV1U9=","name":"","email":"gilfoyle.bertram@hooli.com","groups":["Pied_Piper_Admin"],"admin_group_member":true}
```

We have multiple Harbor instances, one for prod (v2.2.1-b0d63082) and one for test (v2.4.1-c4b06d79). Test has a newer version and is not working. Prod works fine with equal settings. We're hesitating to update prod as this issue would have a very high impact.

We're using an OIDC logging provider with the following settings:*

| Auth Mode            | OIDC                                       |
|----------------------|--------------------------------------------|
| OIDC Provider Name   | HOOLI-CORP                                 |
| OIDC Endpoint        | https://login.hooli.com/adfs               |
| OIDC Client ID       | REDACTED                                   |
| OIDC Client Secret   | REDACTED                                   |
| Group Claim Name     | group                                      |
| OIDC Admin Group     | Pied_Piper_Admin                           |
| OIDC Scope           | email,offline_access,openid,profile,groups |
| Verify Certificate   | TRUE                                       |
| Automatic onboarding | TRUE                                       |
| Username Claim       | email                                      |  

\* _Sensitive info has been redacted_

## Expected behaviour
OIDC login should succeed with auto-onboard enabled, as the user info string contains the email of the logged in user. Instead the user instantly gets the following error after successfully logging in at our OIDC provider and thus is unable to access Harbor:
```
{"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
```
This seems to correspond with error log below in harbor-core:
```
[DEBUG] [/pkg/oidc/helper.go:209]: Raw ID token for verification: REDACTED
[WARNING] [/pkg/oidc/helper.go:271]: Failed to get userInfo by calling remote userinfo endpoint, error: 401 Unauthorized:  
[DEBUG] [/pkg/oidc/helper.go:281]: Fall back to user data from ID token.
[DEBUG] [/core/controllers/oidc.go:132]: Doing automatic onboarding
[ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"unable to recover username for auto onboard, username claim: email"}]} 
```
I checked the Raw JWT token by decoding it back to JSON, which results in:
```json
{
  "aud": "REDACTED",
  "iss": "https://login.hooli.com/adfs",
  "iat": 1640792352,
  "exp": 1640795952,
  "auth_time": 1640792238,
  "mfa_auth_time": 1640792250,
  "sub": "REDACTED",
  "upn": "gilfoyle.bertram@hooli.com",
  "unique_name": "HOOLI\\gilfoyle001",
  "pwd_exp": "4902352",
  "sid": "REDACTED",
  "email": "gilfoyle.bertram@hooli.com",
  "group": [
    "Pied_Piper_Admin"
  ],
  "apptype": "Confidential",
  "appid": "REDACTED",
  "authmethod": "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport",
  "ver": "1.0",
  "scp": "profile openid offline_access email"
}
```
To me this looks to be complete, as it contains both the neccessary username claim (email) and list of group(s). 

## Steps to reproduce
1. Enable login through OIDC and ADFS provider
1. Enable auto-onboard
1. Try to login as OIDC-user  


## Workaround
When disabling auto-onboard, OIDC login does work and the User info string seem to contain the necessary info. 
```
[DEBUG] [/server/middleware/security/unauthorized.go:28][requestID="c29tZVRoaW5nUmFuRG9tRk9SZ2lUSVNTdWU="]: an unauthorized security context generated for request POST /c/oidc/onboard
[DEBUG] [/core/controllers/oidc.go:225]: User info string: {"iss":"https://login.hooli.com/adfs","sub":"YzI5dFpWUm9hVzVuVW1GdVJHOXRSazlTWjJsVVNWTlRkV1U9=","name":"","email":"gilfoyle.bertram@hooli.com","groups":["Pied_Piper_Admin"],"admin_group_member":true}
[DEBUG] [/core/controllers/oidc.go:193]: User created: gilfoyle.bertram@hooli.com
```

## Versions
Our instance is installed using the latest Helm Chart (2.4.1) onto our Kubernetes cluster:
* Harbor: 2.4.1
* Docker engine: 19.3.12
* Helm Version: 3.7.2

## Additional context:
I think information above shoul be sufficient to diagnose the issue, but if not; please let me know and I'll provide everything that's neccessary.

--- a/src/pkg/oidc/helper.go
+++ b/src/pkg/oidc/helper.go
@@ -355,7 +355,10 @@ func userInfoFromClaims(c claimsProvider, setting cfgModels.OIDCSetting) (*UserI
 		}
 
 		if username, ok := allClaims[setting.UserClaim].(string); ok {
-			res.autoOnboardUsername = username
+			// res.Username and autoOnboardUsername both need to be set to create a fallback when mergeUserInfo has not been successfully called.
+			// This can for example occur when remote fails and only a local token is available for onboarding.
+			// Otherwise the onboard flow only has a fallback when "name" is set in the token, which is not always the case as a custom Username Claim could be configured.
+			res.autoOnboardUsername, res.Username = username, username
 		} else {
 			log.Warningf("OIDC. Failed to recover Username from claim. Claim '%s' is invalid or not a string", setting.UserClaim)
 		}

--- a/src/pkg/oidc/helper_test.go
+++ b/src/pkg/oidc/helper_test.go
@@ -284,7 +284,7 @@ func TestUserInfoFromClaims(t *testing.T) {
 				Issuer:              "",
 				Subject:             "",
 				autoOnboardUsername: "airadier@gmail.com",
-				Username:            "Alvaro",
+				Username:            "airadier@gmail.com", // Set Username based on configured UserClaim
 				Email:               "airadier@gmail.com",
 				Groups:              []string{},
 				hasGroupClaim:       false,

